pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,889 @@
+func init() {
+	newTestServer()
+	generateTestCertificate()
+}
+
+// initMocks is function to initialize mocks
+func initMocks(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockAdapter = edgehub.NewMockAdapter(mockCtrl)
+}
+
+//newTestServer() starts a fake server for testing
+func newTestServer() {
+	flag := true
+	TestServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.Contains(r.RequestURI, "/proper_request"):
+			w.Write([]byte("ws://127.0.0.1:20000"))
+		case strings.Contains(r.RequestURI, "/bad_request"):
+			w.WriteHeader(http.StatusBadRequest)
+		case strings.Contains(r.RequestURI, "/wrong_url"):
+			if flag {
+				w.WriteHeader(http.StatusNotFound)
+				flag = false
+			} else {
+				w.Write([]byte("ws://127.0.0.1:20000"))
+			}
+		}
+	}))
+}
+
+//generateTestCertificate() generates fake certificates and stores them at a temporary location (/tmp/kubeedge/testData)
+func generateTestCertificate() {
+	template := &x509.Certificate{
+		IsCA:                  true,
+		BasicConstraintsValid: true,
+		SubjectKeyId:          []byte{1, 2, 3},
+		SerialNumber:          big.NewInt(1234),
+		Subject: pkix.Name{
+			Country:      []string{"test"},
+			Organization: []string{"testor"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	// generate private key
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Println(err)
+	}
+	publicKey := &privateKey.PublicKey
+
+	// create a self-signed certificate. template = parent
+	var parent = template
+	cert, err := x509.CreateCertificate(rand.Reader, template, parent, publicKey, privateKey)
+	if err != nil {
+		fmt.Println(err)
+	}
+	err = os.MkdirAll("/tmp/kubeedge/testData", 0777)
+	if err != nil {
+		fmt.Println(err)
+	}
+	pKey := x509.MarshalPKCS1PrivateKey(privateKey)
+
+	// this will create plain text PEM file.
+	certFile, err := os.Create("/tmp/kubeedge/testData/edge.crt")
+	if err != nil {
+		fmt.Println("Error in creation of file")
+	}
+	var pemkey = &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert}
+	pem.Encode(certFile, pemkey)
+	certFile.Close()
+
+	// this will create plain text PEM file.
+	keyfile, err := os.Create("/tmp/kubeedge/testData/edge.key")
+	if err != nil {
+		fmt.Println("Error in creation of file")
+	}
+	var pemkey1 = &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: pKey}
+	pem.Encode(keyfile, pemkey1)
+	keyfile.Close()
+}
+
+//TestNewEdgeHubController() tests whether the EdgeHubController returned is correct or not
+func TestNewEdgeHubController(t *testing.T) {
+	tests := []struct {
+		name             string
+		want             *EdgeHubController
+		controllerConfig config.ControllerConfig
+	}{
+		{"TestNewEdgeHubController: Positive Test Case",
```

Comment:
can the name be more descriptive ? what is the positive test case ? any negative tests also we can write ?